libs/ledger/src/chaincode/smart_contract.cpp

```diff
@@ -495,8 +498,119 @@ Contract::Result SmartContract::InvokeAction(std::string const &name, chain::Tra
+    {
+      error = "Maximum contract-to-contract call depth (" + std::to_string(MAX_C2C_CALL_DEPTH) +
+              ") exceeded";
+
+      return false;
+    }
+
+    chain::Address called_contract_address;
+    if (!chain::Address::Parse(identity, called_contract_address))
+    {
+      error = "Invalid contract address format '" + identity + "'";
+
+      return false;
+    }
+
+    if (call_history.find(called_contract_address) != call_history.end())
+    {
+      error = "Contract-to-contract call cycle detected";
+
+      return false;
+    }
+
+    call_history.insert(called_contract_address);
+
+    decltype(auto) c = context();
+
+    // TODO(WK) charge for reading from storage
+    auto loaded_contract = CreateSmartContract<SmartContract>(called_contract_address, *c.storage);
+    if (loaded_contract == nullptr)
+    {
+      error = "Failed to load contract " +
+              static_cast<std::string>(called_contract_address.display()) + " from storage";
+
+      return false;
+    }
+
+    auto &module = *(loaded_contract->module_);
+
+    vm_modules::ledger::BindBalanceFunction(module, *loaded_contract);
+    vm_modules::ledger::BindTransferFunction(module, *loaded_contract);
+    module.CreateFreeFunction("getContext",
+                              [&loaded_contract](vm::VM *) -> vm_modules::ledger::ContextPtr {
+                                return loaded_contract->context_;
+                              });
+
+    vm::VM vm2{&module};
+    loaded_contract->context_ =
+        vm_modules::ledger::Context::Factory(&vm2, tx, context().block_index);
+
+    vm::Compiler compiler{&module};
+
+    std::vector<std::string> errors{};
+
+    vm2.SetIOObserver(vm->GetIOObserver());
+    vm2.SetContractInvocationHandler(contract_invocation_handler);
+    vm2.AttachOutputDevice(fetch::vm::VM::STDOUT, vm->GetOutputDevice(fetch::vm::VM::STDOUT));
+
+    // Ensure the new VM breaks when charge limit is reached
+    auto const reference_charge = vm->GetChargeTotal();
+    vm2.SetChargeLimit(vm->GetChargeLimit());
+    vm2.IncreaseChargeTotal(reference_charge);
+
+    vm::ParameterPack param_pack{vm2.registered_types(), std::move(parameters)};
+
+    ContractContext ctx{c.token_contract, called_contract_address, c.storage, c.state_adapter,
+                        c.block_index};
+    ContractContextAttacher raii{*loaded_contract, ctx};
+    c.state_adapter->PushContext(identity);
+
+    bool const success =
+        vm2.Execute(*loaded_contract->executable(), function.name, error, output, param_pack);
+    if (!success)
+    {
+      std::ostringstream ss;
+      ss << "Execution of function " << function.name << " from contract " << identity
+         << " failed with error \"" << error << "\"";
+      error = ss.str();
+    }
+
+    c.state_adapter->PopContext();
+    vm->IncreaseChargeTotal(vm2.GetChargeTotal() - reference_charge);
+
+    auto it = call_history.find(called_contract_address);
+    call_history.erase(it);
```

Comment:
Think that it is probably safer (and more concise) in the following form:

```suggestion
    call_history.erase(called_contract_address);
```